File: beginner_tutorials/src/add_two_ints_client.cpp
#include "ros/ros.h"
#include "beginner_tutorials/AddTwoInts.h" // 作成したサービス(パッケージ名/サービス名.hの形でinclude)
#include <cstdlib>

/**
 * サービスクライアント(C++)のサンプル
 */

int main(int argc, char **argv)
{
  // ノード名「add_two_ints_client」
  ros::init(argc, argv, "add_two_ints_client");

  // ファイル実行時の引数の数を判定
  if (argc != 3)
  {
    ROS_INFO("usage: add_two_ints_client X Y");
    return 1;
  }

  ros::NodeHandle n;

  // サービスクライアントの定義
  // サービス名「add_two_ints」を呼び出す
  ros::ServiceClient client = n.serviceClient<beginner_tutorials::AddTwoInts>("add_two_ints");

  // サービスに送るリクエストのサービスを作成(ファイル実行時の引数を設定)
  beginner_tutorials::AddTwoInts srv;
  srv.request.a = atoll(argv[1]);
  srv.request.b = atoll(argv[2]);

  // サービスの呼び出し
  // サービスの処理が終了するまで待機になる(正常終了ならtrueが返ってくる)
  if (client.call(srv))
  {
    // 正常終了の場合
    ROS_INFO("Sum: %ld", (long int)srv.response.sum);
  }
  else
  {
    // 異常終了の場合
    ROS_ERROR("Failed to call service add_two_ints");
    return 1;
  }

  return 0;
}

File: beginner_tutorials/scripts/add_two_ints_client.py
#!/usr/bin/python3
import sys
import rospy
from beginner_tutorials.srv import * # サービスファイルすべて読み込み

"""
サービスクライアント(python)のサンプル
"""

# サービスを実行する関数
def add_two_ints_client(x, y):
  # サービス名「add_two_ints_python」が使えるようになるまで待機
  rospy.wait_for_service('add_two_ints_python')

  try:
    # サービスのハンドラを作成
    add_two_ints = rospy.ServiceProxy('add_two_ints_python', AddTwoInts)
    # サービスの実行
    resp1 = add_two_ints(x, y)
    return resp1.sum
  except rospy.ServiceException as e:
    print("Service call failed: %s" % e)


# 引数がおかしい場合に表示する関数
def usage():
    return "%s [x y]"%sys.argv[0]


if __name__ == "__main__":
  # ファイル実行時の引数判定
  if len(sys.argv) == 3:
    # 引数の数が正しい場合
    x = int(sys.argv[1])
    y = int(sys.argv[2])
    print("Requesting %s+%s"%(x, y))
    print("%s + %s = %s"%(x, y, add_two_ints_client(x, y)))
  else:
    # 引数の数がおかしい場合
    print(usage())
    sys.exit(1)
